### PR TITLE
Fastfile parser/linter best-practice-analyzer

### PR DESCRIPTION
**Pull request by hjanuschka**
as discussed in: https://github.com/fastlane/fastlane/issues/7368

bildschirmfoto 2016-12-13 um 14 49 44

### CODE REVIEW

**Comment by KrauseFx** [comment]
❤

File: fastlane/README.md [hunk 1]
```diff
@@ -86,7 +86,7 @@ fastlane release
 :rocket: | Saves you **hours** for every app update you release
 :pencil2: | Very flexible configuration using a fully customisable `Fastfile`
 :mountain_cableway: | Implement a fully working Continuous Delivery process
-:ghost: | [Jenkins Integration](https://docs.fastlane.tools/best-practices/continuous-integration/#jenkins-integration): Show the output directly in the Jenkins test results
+:ghost: | [Jenkins Integration](https://github.com/fastlane/fastlane/blob/master/fastlane/docs/Jenkins.md): Show the output directly in the Jenkins test results
```
Comment (KrauseFx) [inline fastlane/README.md]: Should this change be part of the PR?

Comment (KrauseFx) [inline fastlane/README.md]: The link to docs.fastlane.tools seems correct to me 👍

File: fastlane/lib/fastlane/commands_generator.rb [hunk 2]
```diff
@@ -86,6 +86,29 @@ def run
 
       always_trace!
 
+      command "validate-fastfile" do |c|
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: Should we call this `fastlane lint` instead?

File: fastlane/lib/fastlane/commands_generator.rb [hunk 3]
```diff
@@ -86,6 +86,29 @@ def run
...
+        c.syntax = 'fastlane validate-fastfile /path/to/Fastfile'
+        c.option '--platforms STRING', String, 'Comma seperated list of platforms (ios, mac, android) of actions to validate (others will be skipped)'
+        c.action do |args, options|
+          if File.exist?(args.first)
+
+            plat = []
+            plat = options.platforms.split(",") if options.platforms
+            fl_parser = FastfileParser.new(content: File.read(args.first), filepath: args.first, name: "Fastfile", platforms: plat)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: This should look for the default Fastfile by default. Use `FastlaneFolder.fastfile_path` and make sure to test it with and without a Fastfile

File: fastlane/lib/fastlane/commands_generator.rb [hunk 4]
```diff
@@ -86,6 +86,29 @@ def run
...
+            table = fl_parser.analyze
+            puts table
+            counters = fl_parser.counters
+          else
+            raise "File not found"
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: This should be `Fastfile at path not found: #{File.expand_path(path_here)}`

This it's very clear on where _fastlane_ looked

File: fastlane/lib/fastlane/commands_generator.rb [hunk 5]
```diff
@@ -86,6 +86,29 @@ def run
 
       always_trace!
 
+      command "validate-fastfile" do |c|
+        c.syntax = 'fastlane validate-fastfile /path/to/Fastfile'
+        c.option '--platforms STRING', String, 'Comma seperated list of platforms (ios, mac, android) of actions to validate (others will be skipped)'
+        c.action do |args, options|
+          if File.exist?(args.first)
+
+            plat = []
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: Let's write this out to be `platforms`

File: fastlane/lib/fastlane/commands_generator.rb (hunk 3 above)
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: Move detecting of the path out

File: fastlane/lib/fastlane/commands_generator.rb [hunk 6]
```diff
@@ -86,6 +86,29 @@ def run
...
+            puts table
+            counters = fl_parser.counters
+          else
+            raise "File not found"
+          end
+
+          if counters[:all] == 0
+            puts "File succesfully validated".green
+          else
+            puts "Found: Errors/Deprecations/Info: #{counters[:errors].to_s.red}/#{counters[:deprecations].to_s.yellow}/#{counters[:infos].to_s.blue}"
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/commands_generator.rb]: `Fastfile has some issues: ...`

File: fastlane/lib/fastlane/fast_file.rb [hunk 7]
```diff
@@ -217,9 +217,8 @@ def import(path = nil)
     # @param url [String] The git URL to clone the repository from
     # @param branch [String] The branch to checkout in the repository
     # @param path [String] The path to the Fastfile
-    def import_from_git(url: nil, branch: 'HEAD', path: 'fastlane/Fastfile')
+    def import_from_git(url: nil, branch: 'HEAD', path: 'fastlane/Fastfile', return_file: false)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fast_file.rb]: 1 line above, make sure to add a comment

File: fastlane/lib/fastlane/fast_file.rb [hunk 8]
```diff
@@ -250,7 +249,7 @@ def import_from_git(url: nil, branch: 'HEAD', path: 'fastlane/Fastfile')
           rescue
             # We don't care about a failure here, as local actions are optional
           end
-
+        return clone_folder if return_file
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fast_file.rb]: Wrong spacing here, make sure it's aligned with `import`

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 9]
```diff
@@ -0,0 +1,324 @@
+require 'parser'
+require 'parser/current'
+require 'pp'
+require "terminal-table"
+module Fastlane
+  class FastfileParser
+    attr_accessor :original_action
+
+    def to_s
+      "REMOVE_IT"
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fastfile_parser.rb]: Hehe :)

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+      a = find_alias(@original_action.to_sym) unless a
+      return return_data unless a
+
+      # Get the importet file
+      out_channel = StringIO.new
+      $stdout = out_channel
+      $stderr = out_channel
+      if @original_action.to_s == "import_from_git"
+        begin
+        fl = Fastlane::FastFile.new
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fastfile_parser.rb]: Wrong spacing

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        Fastlane::Actions.load_external_actions(actions_path) if File.directory?(actions_path)
+
+        icontent = File.read("#{path}/fastlane/Fastfile")
+        fl_parser = FastfileParser.new(content: icontent, filepath: @dirname, name: "Imported at #{@line_number}", change_dir: false, platforms: @platforms)
+        fl_parser.analyze
+        @content << "#########################################\n"
+        @content << "# BEGIN import_from_git at: #{@line_number} BEGIN\n"
+        @content << "#########################################\n"
+        @content << icontent
+        @content << "#########################################\n"
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fastfile_parser.rb]: That's a lot of code just to import from Git. How about we don't support this here, but just show a message that only the local Fastfile will be linted, with instructions on how to also lint/test the Fastfile that should be imported?

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+          $stderr = STDERR
+          return return_data
+        end
+      end
+
+      options_avail = a.available_options
+      return return_data if options_avail.nil?
+
+      # Validate Options
+      if options_avail.length > 0 && options_avail.first.kind_of?(FastlaneCore::ConfigItem)
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fastfile_parser.rb]: Please don't use short forms of words, use `options_available`

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+          src_code = copy_code
+          @original_action = child.children[1].to_s
+          dropper = '
+
+          '
+          UI.important(src_code) if $verbose
+          begin
+            Dir.chdir(@dirname) do
+              UI.important "CHDIR to: #{@dirname}" if $verbose
+              result = eval(dropper + src_code) # rubocop:disable Lint/Eval
```
Comment (KrauseFx) [inline fastlane/lib/fastlane/fastfile_parser.rb]: Why do we have to `eval` code here?

**Review by KrauseFx — DISMISSED** [review]

**Comment by KrauseFx** [comment]
Hey @hjanuschka, I love the outcome, this is some really cool things. Did I understand correctly that this doesn't eval the Fastfile, but uses a Ruby code parser instead to iterate over the different checks? I also noticed that one eval code is inside. Also some more notes.

I'd love to have this feature, however I'm also a little worried about maintainability.

**Comment by hjanuschka** [comment]
@KrauseFx  thx for the review, i ll address the issues later (still a bit lazzy out of vacation ) 
but basically - yes it uses ruby parser to parse Fastfile - extracting lanenames actionnames and so on.

but to further validate options from various actions, it extracts the action names - replaces it with a `dummy` method name and evals it.

i am not really worried about this eval, in that special case.
about the maintainability you are right, we would atleast need a few more eyes on it.

File: fastlane/README.md (hunk 1 above)
Comment (hjanuschka) [inline fastlane/README.md]: done - not sure why this happened :(

File: fastlane/lib/fastlane/commands_generator.rb (hunk 2 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: done

File: fastlane/lib/fastlane/commands_generator.rb (hunk 5 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: done

File: fastlane/lib/fastlane/commands_generator.rb (hunk 3 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: done - awesome, did'nt know about the FastlaneFolder

File: fastlane/lib/fastlane/commands_generator.rb (hunk 4 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: done

File: fastlane/lib/fastlane/commands_generator.rb (hunk 6 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: done.

bildschirmfoto 2017-01-08 um 20 12 49

File: fastlane/lib/fastlane/fast_file.rb (hunk 7 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fast_file.rb]: done

File: fastlane/lib/fastlane/fast_file.rb (hunk 8 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fast_file.rb]: done

File: fastlane/lib/fastlane/fastfile_parser.rb (hunk 9 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: thats a placeholder for variables 👍  like if user has something like this:


```
action(param: "#{some_ruby_thing} asdfg")
```

those will be replaced by a FastfileParser class - and via `to_s`  replaced with "REMOVE_IT".

File: fastlane/lib/fastlane/fastfile_parser.rb (hunk 10 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: done.

File: fastlane/lib/fastlane/fastfile_parser.rb (hunk 11 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: well - atleast for me with my setup the linting without support for `import_from_git` is useless - as we only have a 4 line Fastfile per project (3 variables  - and the call to `import_from_git`)

i know that it feels a bit clunky but i think its worth the benefit from supporting `import_from_git`


also i think describing/finding a way to `lint`  git fastfiles is hard. (involves, clone, chdir, ensure appfile and stuff exists, lint)

File: fastlane/lib/fastlane/fastfile_parser.rb (hunk 12 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: thx

File: fastlane/lib/fastlane/fastfile_parser.rb (hunk 13 above)
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: see last comment, mostly because of the need to eval the source code returned by the parser.

**Review by hjanuschka — COMMENTED** [review]
feedback addressed, and `sh` should validate now again.

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+    def redacted
+      rc = ""
+      rc = @content
+      self.class.secrets.each do |s|
+        rc.gsub!(s, "#" * s.length)
+      end
+      rc
+    end
+
+    # rubocop:disable PerceivedComplexity
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: @KrauseFx  any idea on how to fix it? - i am not sure on how to split the whole thing up without reaching the `PerceivedComplexity`

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 15]
```diff
@@ -0,0 +1,334 @@
+require 'parser'
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: @KrauseFx what is this?

**Comment by hjanuschka** [comment]
added support for the newly merged #7609 

bildschirmfoto 2017-01-17 um 19 50 36


i needed to force push, to get the current master into this PR.

this is ready for another review!

**Review by mfurtak — CHANGES_REQUESTED** [review]
Hey @hjanuschka - I've left a number of small comments about specific things. However, on the larger scale, I wouldn't be in favor of merging this as it currently stands.

Rubocop is not always right, but in this case I agree with the `PerceivedComplexity` warning about `fake_action`. That method is quite large, and does a lot of different things. The `FastfileParser` class as a whole has many responsibilities, and I think it would benefit from being broken up into smaller classes.

So, without trying to prescribe a particular solution, I might suggest having a look through [rubocop](https://github.com/bbatsov/rubocop)'s code, and see how it is organized. One of the things that "feels good" about the Rubocop code is having each check defined by its own class/file. But again, that's just an example, not a requirement :smile:

**Comment by hjanuschka** [comment]
@mfurtak thx for the feedback, i am going to adress the inline comments, and try to split it up so rubocop does not require exceptions (except the eval).

File: fastlane/lib/fastlane/commands_generator.rb [hunk 16]
```diff
@@ -95,6 +95,31 @@ def run
 
       always_trace!
 
+      command "lint" do |c|
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: it initially was 'validate-fastlane' - which i didnt get to work with symbol. -> fixed

File: fastlane/lib/fastlane/commands_generator.rb [hunk 17]
```diff
@@ -95,6 +95,31 @@ def run
...
+        c.syntax = 'fastlane validate-fastfile /path/to/Fastfile'
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: thx

File: fastlane/lib/fastlane/commands_generator.rb [hunk 18]
```diff
@@ -95,6 +95,31 @@ def run
...
+        c.option '--platforms STRING', String, 'Comma seperated list of platforms (ios, mac, android) of actions to validate (others will be skipped)'
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: awesome thats nice, didn't know! 🚀

File: fastlane/lib/fastlane/commands_generator.rb [hunk 19]
```diff
@@ -95,6 +95,31 @@ def run
...
+        c.action do |args, options|
+          file_path = args.first
+          file_path ||= FastlaneCore::FastlaneFolder.fastfile_path
+          if File.exist?(file_path)
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: thx

File: fastlane/lib/fastlane/commands_generator.rb [hunk 20]
```diff
@@ -95,6 +95,31 @@ def run
...
+          end
+
+          if counters[:all] == 0
+            puts "File succesfully validated".green
+          else
+            puts "Fastfile has some issues: #{counters[:errors].to_s.red} Errors, #{counters[:deprecations].to_s.yellow} Deprecations #{counters[:infos].to_s.blue} Infos"
+          end
+        end
+      end
       command :trigger do |c|
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/commands_generator.rb]: thx

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 21]
```diff
@@ -0,0 +1,347 @@
...
+      rc = @content
+      self.class.secrets.each do |s|
+        rc.gsub!(s, "#" * s.length)
+      end
+      rc
+    end
+
+    # rubocop:disable PerceivedComplexity
+    def fake_action(*args)
+      # Suppress UI output
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: removed it

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 22]
```diff
@@ -0,0 +1,347 @@
...
+
+      return_data = { args: args.first }
+      return return_data if @original_action.nil?
+      UI.important("ACTION: #{@original_action.inspect} 1") if $verbose
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: removed all the unnecessary debug outputs

File: fastlane/lib/fastlane/fastfile_parser.rb [hunk 23]
```diff
@@ -0,0 +1,347 @@
...
+          if a.respond_to?(:deprecated_notes)
+            deprecated_notes = a.deprecated_notes
+          end
+          lines << { state: :deprecated, line: @line_number, msg: "Action `#{@original_action}` is deprecated!\n #{deprecated_notes}" }
+        end
+      end
+
+      return return_data if args.length <= 0
+
+      # Get the importet file
```
Comment (hjanuschka) [inline fastlane/lib/fastlane/fastfile_parser.rb]: thx

**Review by hjanuschka — COMMENTED** [review]
@mfurtak  thx i pushed a new commit with feedback addressed and rubocop in a good mood without `disable`  rule.

**Comment by mfurtak** [comment]
I think this is an interesting idea, but probably one that should be explored outside of _fastlane_ itself. 👍

**Pull request closed without merging**